### ISSUE TITLE
Introduce the "set anchor" feature / Bookmarks

### ISSUE THREAD

**Issue by kleinph** [ckeditor/ckeditor5#1944]
## Is this a bug report or feature request? (choose one)
🆕 Feature request

Create a feature that allows the user to insert an anchor to the document, just like in CKEditor 4.
<img width="317" alt="Screenshot 2022-05-02 at 09 37 37" src="https://user-images.githubusercontent.com/5314687/166200645-143aea05-f17e-497c-ab81-d43efca7acd7.png">

<img width="466" alt="Screenshot 2022-05-02 at 09 38 10" src="https://user-images.githubusercontent.com/5314687/166200655-ee8644e0-10c1-440d-9d1d-4ef3cdd6a44f.png">

------

If you'd like to see this feature implemented, add 👍 to this post.

**Comment by Mgsy** [comment]
Hi, thanks for the report. I've converted your ticket to a feature request.

**Comment by meecect** [comment]
I've been trying to write a plugin using the downcast converter to automatically add an anchor tags to heading tags.

I can't get it to work...in the interim before this feature request is finished, could someone help with the code requried for this?  I have followed the docs and have been able to add classses, set attributes, etc, but wheh I try to insert elements, it doesn't seem to work.

Could you give an example of how one could add the linkHref attribute to the text inside a heading tag in a downcast action?

Or, alternatively, (perhaps better), could one create a blank text element with an href attribute and append (or prepend)  it to the heading1 element in a downcast action?

**Comment by kaleu50** [comment]
any news for this feature?

**Comment by aborkhovych** [comment]
+1 to this feature request

**Comment by 523299338** [comment]
+2 to this feature request

**Comment by wimleers** [comment]
Drupal would also benefit from this feature: https://www.drupal.org/project/ckeditor5/issues/3224256.

**Comment by hools18** [comment]
+1 to this feature request

**Comment by PTWvL** [comment]
Google brought me here. +1 for this request

**Comment by dpf-dd** [comment]
#meetoo <3

**Comment by vlyalkobixal** [comment]
+1 for this too

**Comment by shootlens** [comment]
+4 for this feature request

**Comment by Barik85** [comment]
+1 for this feature request

**Comment by cmak9149** [comment]
+1

**Comment by venmathiraj** [comment]
+1

**Comment by ferrlen** [comment]
Is there any progress being made on this issue? I'd like to use CKEditor 5 in my app, but this is a dealbreaker.

It's not particularly hard to implement this feature. I thought of creating a plugin to add anchors, but this wouldn't make much of a difference because links in CKEditor seem to always open in a new window, which is not the behaviour I want. I just want the user to be able to scroll to the desired point in the text.

**Comment by javadbat** [comment]
+1

**Comment by ndanvers** [comment]
+1 to this.. It's quite a shame that this is not available in CKE5, when, to my knowledge, its available in CKE4.

**Comment by hponcet** [comment]
+1 Can't wait to see this feature come out!

**Comment by theMusician** [comment]
+1

**Comment by taoyu65** [comment]
+1

**Comment by artemboyko43** [comment]
+1

**Comment by realysy** [comment]
seems like this is a forgotten issue...

**Comment by BryanDeNijs** [comment]
+1

**Comment by bob-hinrichs** [comment]
+1

**Comment by bennydv** [comment]
+1

**Comment by helloandrewpaul** [comment]
+1

**Comment by loopy3025** [comment]
+1

**Comment by kitano-takuya1984** [comment]
+1

**Comment by ssougnez** [comment]
+1

**Comment by mayelaatf** [comment]
+1

**Comment by hkirsman** [comment]
+1

**Comment by ndanvers** [comment]
Bump - It's frankly ridiculous that CKeditor 5 is still not able to provide this functionality - And for more than FOUR YEARS we've had nothing but crickets.

This is but one reason as to why I wouldn't choose CKeditor 5 in any project I am starting. 

What a disaster of a repo.

**Comment by raq007** [comment]
+1 to this feature request

**Comment by marcbertone** [comment]
+1

**Comment by raq007** [comment]
@Reinmar thank you for the above update. 

By any chance do you think this feature will be available by End of this year ? We are migrating to CKE5 under a time constraint so would be very helpful to know.

Thanks!

**Comment by queenielow** [comment]
+1

**Comment by amandine-ls** [comment]
+1

**Comment by amandine-ls** [comment]
+1

**Comment by Kwaadpepper** [comment]
+1 Really need this feature. If you have time for it, it would help lots of people.

**Comment by amandine-ls** [comment]
+1

**Comment by skyyip-theorigo** [comment]
+1

**Comment by Kwaadpepper** [comment]
If anyone is interested, I found that people has ported a Creditor plugin for Drupal usage. I have successfully used it in my custom build.

https://github.com/northernco/ckeditor5-anchor-drupal

**Comment by dianahub** [comment]
> If anyone is interested, I found that people has ported a Creditor plugin for Drupal usage. I have successfully used it in my custom build.
> 
> https://github.com/northernco/ckeditor5-anchor-drupal

Hi Kwaadpepper can you explain how you add this plugin. ?  I tried putting it into the plugin directory and nothing new shows up.

**Comment by dianahub** [comment]
this just came out https://www.drupal.org/project/anchor_link

**Comment by dianahub** [comment]
The library in github is missing the build directory, i found the entire library here , under package https://registry.npmjs.org/@northernco/ckeditor5-anchor-drupal/-/ckeditor5-anchor-drupal-0.4.0.tgz

**Comment by Kwaadpepper** [comment]
> > If anyone is interested, I found that people has ported a Creditor plugin for Drupal usage. I have successfully used it in my custom build.
> > 
> > https://github.com/northernco/ckeditor5-anchor-drupal
> 
> Hi Kwaadpepper can you explain how you add this plugin. ?  I tried putting it into the plugin directory and nothing new shows up.

In order to use the plugin you shall follow instructions to create your own build

https://ckeditor.com/docs/ckeditor5/latest/installation/getting-started/quick-start-other.html#building-the-editor-from-source

You will need a terminal of course with Npm and NodeJs.

**Comment by Kwaadpepper** [comment]
> The library in github is missing the build directory, i found the entire library here , under package https://registry.npmjs.org/@northernco/ckeditor5-anchor-drupal/-/ckeditor5-anchor-drupal-0.4.0.tgz

I indeed added it from Npm in my custom build. As a matter effect you can always search on Npm packages that has the same name as the GitHub repo, this is usually the case.
Then add it to your project `npm i -D the-npm-package`

**Comment by skyyip-theorigo** [comment]
+1 to this feature request

**Comment by amandine-ls** [comment]
+1

**Comment by NvanWeeghel** [comment]
+1

**Comment by ttrench** [comment]
Any news on this feature request? Would love to have it available in CKE5.

**Comment by amandine-ls** [comment]
+1

**Comment by goPOSTALman** [comment]
Still waiting on this to be added since 2020... wooo!

**Comment by dwandeljanicka** [comment]
Hi! We have started working on anchors and we would like to know your opinion to chose the most useful solution.

**How should the anchor be displayed in the editor? Vote with emoji:**

:rocket: Inline (as it was in CKEditor 4) ![Zrzut ekranu 2024-07-22 o 14 20 01](https://github.com/user-attachments/assets/d4b0a485-aa3e-4005-93d3-1a3ece68c69f)

:tada: Out of the text editing area ![Zrzut ekranu 2024-08-7 o 09 19 42](https://github.com/user-attachments/assets/2f38ee12-747c-4e42-95a8-0a7920d14374)
![Zrzut ekranu 2024-08-7 o 09 22 32](https://github.com/user-attachments/assets/c67852bb-57ed-4160-9ef4-543b6835eda3)

:eyes: I don't need to see the anchors in the editor; it is enough that I can add links to them

**Comment by abigail1309** [comment]
🚀

Abigail Taylor

Product Manager

━━

t: +44 (0)20 3970 3403

e: ***@***.***


[image: Passle Logo] <https://www.passle.net/>

The Thought Leadership Platform for Professional Services Firms

Subscribe to the Passle CMO Series Podcast
<https://pssle.co/passle-insights/s/2f42> | blog.passle.net


UK office | Passle Limited, 1a Walton Crescent, Oxford OX1 2JG, UK
US office | Passle Inc, 222 Severn Avenue, Suite 50, Building 10-3,
Annapolis MD 21403, USA

**Comment by raq007** [comment]
🚀

**Comment by joegl** [comment]
🚀

**Comment by justcaldwell** [comment]
🚀

**Comment by NIWm** [comment]
🚀

**Comment by SimonChabrier** [comment]
🚀

**Comment by SimonChabrier** [comment]
When will the feature be available please ?

**Comment by Witoso** [comment]
We are working on this initiative right now. Most likely it will be done in 3 stages, though released together. I will update this comment with the links to more details as we go.

*   #17063 
*   \[Part 2\] Linking experience to bookmarks
*   \[Part 3\] Polishing and integrating across different features

**Comment by Witoso** [comment]
The work on bookmarks progresses quite nicely, and we decided to make a release of it without waiting for linking experience improvements ([see scope in the comment](https://github.com/ckeditor/ckeditor5/issues/1944#issuecomment-2338002012)) That way, you will be able to use the plugin earlier. TBD still on the exact release date.

But there's a second part that we will be focusing on immediately afterward: linking improvements and linking to bookmarks (#17230). I added a few UI prototypes how improved linking balloon may look like, and which areas will change. Please don't hesitate to leave your comments on the issue 😊

**Comment by Witoso** [comment]
Hey, we've finished the main work. This is planned for release but for the next iteration. We've just started testing phase for a bug fix release (planned for beginning of November). This is scheduled for the next one (~beginning of December).

Meanwhile, we continue to work on linking improvements and linking to bookmarks (#17230).

**Comment by dalin-** [comment]
I love getting a native implementation into CKEditor5.  I have one suggestion, but I've never posted something in this community, so I have no idea how you do things.  

If the new implementation matches what's shown in the screenshots in the OP description, then there's some UX issues that you might consider addressing before this goes out.  The main issue is that you've got to know a lot about the underlying mechanics of anchors before you're able to use this thing. Most users on most sites won't know enough to be able to make use of this. 

Here's a suggestion for the pop-up window.  

Feel free to run with this, turn it into a new issue, or whatever fits best around here. 

-----------------------

Use an anchor to link to a section within a page (not just the top):

Anchor name
⊏ニニニニニニニニニニ⊐ ✓ ⨯

**To link to this anchor from within the current page:** Don't use the full URL. Instead use a number sign followed by the name of the anchor.
#[anchor-name] ⎘
**From another page:** Use the full URL of this page followed by a number sign and the name of the anchor.

**Comment by Witoso** [comment]
@dalin- thanks for the feedback!

We will release the bookmarks without those, but I will make sure to mention in the docs that it's a feature v1 without all final touches. The final UI is connected with the #17230 where the list of bookmarks to link to will be automatically prefilled. We also will provide integrators to create their lists (bookmarks on other pages, for example).

I believe that informative descriptions should be the last resort for intuitive user interfaces, as they make the visuals bloated and the text has always the possibility of being misinterpreted. We also have tons of different use cases that the editor is used in, and we need to be careful about writing copy that may not be relevant for some of those.
